### PR TITLE
Add MaxText-Omni POC README, improve testing, and fix checkpoint handling

### PR DESCRIPTION
**Pull request by subawocit**
# Description

This PR added comprehensive pipeline documentation. It also fixed the checkpoint saving and restoration logic upon recent upstream maxtext checkpointing changes, and improved overall code hygiene and efficiency.

# Files
- **`README.md`: Adds guide for the entire pipeline.**
- `stitch_checkpoint.py`: Fixes checkpoint saving logic upon recent upstream maxtext checkpointing changes. Adds improvements:
  - Evaluates `ocp.checkpoint_utils.construct_restore_args(subtrees_abstract)` once upfront instead of traversals in each conditional branch.
  - Refactors `_assemble` into a dict comprehension using `restored_val.get(sub_module_name, fresh_weights)`, reducing dictionary key lookups from 2M lookups to M.
- `stitch_checkpoint_test.py`: Fixes checkpoint restoration logic upon recent upstream maxtext checkpointing changes and standardizes on `BASE_OUTPUT_DIRECTORY`. Adds improvements:
  - Replaces recursive DFS in `_get_vision_layer_count` with an explicit LIFO list stack, removing recursion limit risks on deeply nested checkpoints in large-size models.
- `compare_sft_checkpoint_test.py`: Uses `jax.device_put` to align TPU sharding for tensor difference calculation.
- `decode_omni.py` & `train_sft_omni.py`: Code clean up.

# Tests

### 1. End-to-End Pipeline Run (`maxtext_omni_pipeline_e2e.sh`)
```bash
export BASE_OUTPUT_DIRECTORY="gs://yuchenhou-maxtext-logs/e2e_improve"
export HF_TOKEN="<HF-TOKEN>"
export TRAIN_STEPS=20
export EVAL_NUM_EXAMPLES=10
export SCAN_LAYERS=true

bash src/maxtext/experimental/omni_poc/maxtext_omni_pipeline_e2e.sh
```
**Result:**
```
Omni End-to-End Pipeline Complete!
  Stitched Base:       gs://yuchenhou-maxtext-logs/e2e_improve/omni_checkpoints/omni_stitched_gemma3-4b_qwen3-4b/0/items
  Pretrain Checkpoint: gs://yuchenhou-maxtext-logs/e2e_improve/pretrain_chartnet/pretrain_chartnet/checkpoints/19/items
  SFT Final Checkpoint: gs://yuchenhou-maxtext-logs/e2e_improve/sft_after_chartnet/sft_chartqa/checkpoints/19/items
  Eval Results Dir:    gs://yuchenhou-maxtext-logs/e2e_improve/sft_after_chartnet/eval
```

### 2. Checkpoint Stitching Unit Test (`stitch_checkpoint_test.py`)
```bash
export BASE_OUTPUT_DIRECTORY="gs://yuchenhou-maxtext-logs/e2e_improve"
export PYTHONPATH=src:${PYTHONPATH:-}

python3 -m unittest src/maxtext/experimental/omni_poc/tests/stitch_checkpoint_test.py
```
**Result:**
```
  - Verified LLM layers count (DecoderBlockType.QWEN3): orig=36 and stitched=36
  - Verified Vision layers count (VisionEncoderBlockType.GEMMA3): orig=27 and stitched=27

  - Vision encoder output logits mean of abs: orig=0.09, stitched=0.09 (Mean abs diff = 0.00e+00, max = 0.00e+00)
  - Token embedder output logits mean of abs: orig=0.02, stitched=0.02 (Mean abs diff = 0.00e+00, max = 0.00e+00)
  - Projector output logits mean of abs:     orig=0.03, stitched=0.03 (Mean abs diff = 4.62e-02, max = 1.74e+00)

OK
```

### 3. Checkpoint Weight & Freezing Invariant Test (`compare_sft_checkpoint_test.py`)
```bash
python3 -m maxtext.experimental.omni_poc.tests.compare_sft_checkpoint_test \
  src/maxtext/experimental/omni_poc/configs/sft-maxtext-omni-gemma3-qwen3.yml \
  --stitched_checkpoint_path=gs://yuchenhou-maxtext-logs/e2e_improve/omni_checkpoints/omni_stitched_gemma3-4b_qwen3-4b/0/items \
  --sft_checkpoint_path=gs://yuchenhou-maxtext-logs/e2e_improve/sft_after_chartnet/sft_chartqa/checkpoints/19/items
```
**Result:**
```
  [PASS] Vision Encoder : 437/437 tensors frozen
  [PASS] Projector MLP  : 6/6 tensors updated (max diff: 0.003758)
  [PASS] LLM Backbone   : 13/13 tensors frozen
```

### 4. Processor Unit Test (`processor_maxtext_omni_test.py`)
```bash
python3 -m unittest src/maxtext/experimental/omni_poc/tests/processor_maxtext_omni_test.py
```
**Result:**
```
OK
```

# Checklist

Before submitting this PR, please make sure (put X in square brackets):
- [X] I have performed a self-review of my code. For an optional AI review, add the `gemini-review` label.
- [X] I have necessary comments in my code, particularly in hard-to-understand areas.
- [X] I have run end-to-end tests and provided workload links above if applicable.
- [X] I have made or will make corresponding changes to the doc if needed, including adding new documentation pages to the relevant Table of Contents (toctree directive) as explained in [our documentation](https://maxtext.readthedocs.io/en/latest/development.html#adding-new-documentation-files).

### CODE REVIEW

**Review by hengtaoguo — APPROVED** [review]
Let's GO!

File: src/maxtext/experimental/omni_poc/README.md [hunk 1]
```diff
@@ -0,0 +1,244 @@
...
+python3 -m maxtext.experimental.omni_poc.utils.stitch_checkpoint \
+  src/maxtext/experimental/omni_poc/maxtext-omni-gemma3-qwen3.yml \
+  hf_access_token=${HF_TOKEN} \
+  vision_load_path=${BASE_OUTPUT_DIRECTORY}/converted/gemma3-4b/0/items \
+  llm_load_path=${BASE_OUTPUT_DIRECTORY}/converted/qwen3-4b/0/items \
+  stitched_output_path=${BASE_OUTPUT_DIRECTORY}/omni_checkpoints/omni_stitched_gemma3-4b_qwen3-4b/0/items
+```
+
+#### Step 3: Stage 1 Alignment Pretraining (ChartNet)
+Pretrains the MLP connector on chart summaries from `ibm-granite/ChartNet`. This step aligns the vision tower with the language model through our customized MLP.
```
Comment (lydhr) [inline src/maxtext/experimental/omni_poc/README.md]: It is slightly misleading that the we did `pre-train` the MLP but via `post-training` (SFT)

File: src/maxtext/experimental/omni_poc/README.md [hunk 2]
```diff
@@ -0,0 +1,244 @@
...
+python3 -m maxtext.experimental.omni_poc.train_sft_omni \
+  src/maxtext/experimental/omni_poc/configs/pretrain-maxtext-omni-gemma3-qwen3-chartnet.yml \
+  load_parameters_path=${BASE_OUTPUT_DIRECTORY}/omni_checkpoints/omni_stitched_gemma3-4b_qwen3-4b/0/items \
+  base_output_directory=${BASE_OUTPUT_DIRECTORY}/pretrain_chartnet \
+  run_name=pretrain_chartnet \
+  hf_access_token=${HF_TOKEN}
+```
+
+#### Step 4: Stage 2 Supervised Fine-Tuning (ChartQA)
+Fine-tunes the projector for visual question answering using `HuggingFaceM4/ChartQA`:
```
Comment (lydhr) [inline src/maxtext/experimental/omni_poc/README.md]: Maybe call it curriculum post-train (SFT), including step 3 and 4.

**Review by lydhr — APPROVED** [review]

File: src/maxtext/experimental/omni_poc/README.md (hunk 1 above)
Comment (subawocit) [reply src/maxtext/experimental/omni_poc/README.md]: That's a very good point, I modified the wording in both READMD.md and e2e files to

`Curriculum Post-Training Stage 1: Projector Alignment` and `Curriculum Post-Training Stage 2: Projector Task Fine-Tuning`
